File: sdk/typescript/tests/e2e-admin.ts
/** E2E: admin — schedulers, DLQ, configs, rate limits, webhooks, monitoring. */

import { assert, assertEq, makeQueue, makeWorker, qname, sleep, test, waitFor } from './harness.ts';

test('admin: scheduler every spawns jobs, single get, remove stops it', async () => {
  const queue = makeQueue('sched');
  const schedulerId = qname('tick');
  try {
    await queue.every(schedulerId, 300, { kind: 'tick' });
    const schedulers = await queue.getJobSchedulers();
    assert(
      schedulers.some((s) => s.name === schedulerId),
      'scheduler listed'
    );
    const single = await queue.getJobScheduler(schedulerId);
    assert(single !== null && single.queue === queue.name, 'single scheduler fetched');
    await waitFor(async () => (await queue.count()) >= 1, 15_000);
    await queue.removeJobScheduler(schedulerId);
    const after = await queue.count();
    await sleep(1000);
    assert((await queue.count()) <= after + 1, 'scheduler stopped firing');
  } finally {
    queue.close();
  }
});

test('admin: cron pattern scheduler + CronGet not-found → null', async () => {
  const queue = makeQueue('cron');
  const schedulerId = qname('daily');
  try {
    await queue.upsertJobScheduler(
      schedulerId,
      { pattern: '0 9 * * *', tz: 'Europe/Rome' },
      { name: 'report', data: { type: 'daily' } }
    );
    const single = await queue.getJobScheduler(schedulerId);
    assert(single !== null, 'scheduler fetched');
    assertEq(single?.schedule, '0 9 * * *', 'cron pattern stored');
    assert(Number(single?.nextRun ?? 0) > 0, 'nextRun computed');
    await queue.removeJobScheduler(schedulerId);
    assertEq(await queue.getJobScheduler(schedulerId), null, 'not found → null');
  } finally {
    queue.close();
  }
});

test('admin: DLQ entries + retryDlq requeues', async () => {
  const queue = makeQueue('dlqr');
  const worker = makeWorker(queue.name, async () => {
    throw new Error('dead');
  });
  try {
    const job = await queue.add('boom', {}, { attempts: 1 });
    await waitFor(async () => (await queue.getDlq()).length === 1);
    await worker.close();
    const entries = await queue.getDlq();
    assert(entries[0].jobId === job.id || entries[0].id === job.id, 'DLQ entry references job');
    const retried = await queue.retryDlq();
    assertEq(retried, 1, 'one retried');
    const state = await queue.getJobState(job.id);
    assert(state === 'waiting' || state === 'prioritized', 'back to waiting');
    assertEq((await queue.getDlq()).length, 0, 'DLQ empty');
  } finally {
    if (!worker.isClosed()) await worker.close();
    queue.close();
  }
});

test('admin: purgeDlq', async () => {
  const queue = makeQueue('dlqp');
  const worker = makeWorker(queue.name, async () => {
    throw new Error('dead');
  });
  try {
    await queue.add('boom', {}, { attempts: 1 });
    await waitFor(async () => (await queue.getDlq()).length === 1);
    await worker.close();
    assertEq(await queue.purgeDlq(), 1, 'purged one');
    assertEq((await queue.getDlq()).length, 0, 'DLQ empty');
  } finally {
    if (!worker.isClosed()) await worker.close();
    queue.close();
  }
});

test('admin: stall + dlq config roundtrip', async () => {
  const queue = makeQueue('cfg');
  try {
    await queue.setStallConfig({ stallInterval: 20_000, maxStalls: 5, gracePeriod: 3000 });
    const stall = await queue.getStallConfig();
    assertEq(Number(stall.stallInterval), 20_000, 'stallInterval');
    assertEq(Number(stall.maxStalls), 5, 'maxStalls');
    await queue.setDlqConfig({ autoRetry: true, maxEntries: 500 });
    const dlq = await queue.getDlqConfig();
    assertEq(Number(dlq.maxEntries), 500, 'dlq maxEntries');
  } finally {
    queue.close();
  }
});

test('admin: rate limit + concurrency accepted', async () => {
  const queue = makeQueue('limits');
  try {
    await queue.setGlobalRateLimit(100);
    await queue.removeGlobalRateLimit();
    await queue.setGlobalConcurrency(2);
    await queue.removeGlobalConcurrency();
  } finally {
    queue.close();
  }
});

test('admin: global concurrency enforced via PULLB', async () => {
  const queue = makeQueue('gconc');
  try {
    await queue.setGlobalConcurrency(1);
    await queue.addBulk(Array.from({ length: 4 }, (_, i) => ({ name: 't', data: { i } })));
    const response = await queue.connection.call({
      cmd: 'PULLB',
      queue: queue.name,
      count: 4,
      owner: 'e2e',
      lockTtl: 30_000,
    });
    const jobs = (response.jobs ?? []) as unknown[];
    assert(jobs.length <= 1, `global concurrency not enforced: got ${jobs.length}`);
    await queue.removeGlobalConcurrency();
  } finally {
    queue.close();
  }
});

test('admin: webhooks CRUD (https URL, SSRF blocks localhost)', async () => {
  const queue = makeQueue('hooks');
  try {
    const added = await queue.addWebhook({
      url: 'https://example.com/bunqueue-hook',
      events: ['job.completed'],
    });
    const listed = await queue.listWebhooks();
    assert(
      listed.some((h) => h.url === 'https://example.com/bunqueue-hook'),
      'webhook listed'
    );
    const id = String(added.id ?? added.webhookId ?? '');
    assert(id.length > 0, 'webhook id resolvable');
    await queue.removeWebhook(id);
    const after = await queue.listWebhooks();
    assert(!after.some((h) => h.id === id), 'webhook removed');
  } finally {
    queue.close();
  }
});

test('admin: stats + metrics + listQueues (string names)', async () => {
  const queue = makeQueue('mon');
  try {
    await queue.add('t', {});
    const stats = await queue.getStats();
    assert('waiting' in stats && 'uptime' in stats, `stats shape: ${Object.keys(stats)}`);
    const metrics = await queue.getMetrics();
    assert('totalPushed' in metrics, `metrics shape: ${Object.keys(metrics)}`);
    const queues = await queue.listQueues();
    assert(queues.includes(queue.name), 'queue name listed as string');
  } finally {
    queue.close();
  }
});

File: sdk/python/tests/e2e_admin.py
"""E2E: admin — schedulers, rate limits, configs, webhooks, monitoring, DLQ."""

from __future__ import annotations

import time

from harness import Server, test, unique_name, wait_until

from bunqueue import Queue, Worker


@test
def scheduler_every_spawns_jobs(server: Server) -> None:
    with Queue(unique_name("cron"), port=server.port) as queue:
        scheduler_id = unique_name("tick")
        queue.every(scheduler_id, 300, {"kind": "tick"})
        try:
            schedulers = queue.get_job_schedulers()
            assert any(s["name"] == scheduler_id for s in schedulers), f"missing: {schedulers}"
            single = queue.get_job_scheduler(scheduler_id)
            assert single is not None and single["queue"] == queue.name
            assert wait_until(lambda: queue.count() >= 1, 15), "cron did not spawn jobs"
        finally:
            queue.remove_job_scheduler(scheduler_id)
        count_after_delete = queue.count()
        time.sleep(1.0)
        assert queue.count() <= count_after_delete + 1, "scheduler kept firing after delete"


@test
def scheduler_cron_pattern(server: Server) -> None:
    with Queue(unique_name("cron2"), port=server.port) as queue:
        scheduler_id = unique_name("daily")
        queue.upsert_job_scheduler(
            scheduler_id,
            {"pattern": "0 9 * * *", "tz": "Europe/Rome"},
            {"name": "report", "data": {"type": "daily"}},
        )
        try:
            single = queue.get_job_scheduler(scheduler_id)
            assert single is not None
            assert single.get("schedule") == "0 9 * * *"
            assert single.get("nextRun", 0) > 0
            assert queue.get_job_schedulers_count() == 1
        finally:
            queue.remove_job_scheduler(scheduler_id)
        assert queue.get_job_scheduler(scheduler_id) is None
        assert queue.get_job_schedulers_count() == 0


@test
def dlq_flow_and_retry(server: Server) -> None:
    with Queue(unique_name("dlq"), port=server.port) as queue:
        job = queue.add("boom", {}, attempts=1)
        worker = Worker(
            queue.name,
            lambda j: (_ for _ in ()).throw(RuntimeError("dead")),
            port=server.port,
            poll_timeout_ms=300,
        )
        worker.start()
        try:
            assert wait_until(lambda: len(queue.get_dlq()) == 1, 15)
        finally:
            worker.close(timeout=10)
        entries = queue.get_dlq()
        assert entries[0].get("jobId") == job.id or entries[0].get("id") == job.id, entries
        retried = queue.retry_dlq()
        assert retried == 1
        assert queue.get_state(job.id) in ("waiting", "prioritized")
        assert len(queue.get_dlq()) == 0


@test
def purge_dlq(server: Server) -> None:
    with Queue(unique_name("dlqp"), port=server.port) as queue:
        queue.add("boom", {}, attempts=1)
        worker = Worker(
            queue.name,
            lambda j: (_ for _ in ()).throw(RuntimeError("dead")),
            port=server.port,
            poll_timeout_ms=300,
        )
        worker.start()
        try:
            assert wait_until(lambda: len(queue.get_dlq()) == 1, 15)
        finally:
            worker.close(timeout=10)
        assert queue.purge_dlq() == 1
        assert len(queue.get_dlq()) == 0


@test
def stall_and_dlq_config(server: Server) -> None:
    with Queue(unique_name("cfg"), port=server.port) as queue:
        queue.set_stall_config({"stallInterval": 20000, "maxStalls": 5, "gracePeriod": 3000})
        config = queue.get_stall_config()
        assert config.get("stallInterval") == 20000, config
        assert config.get("maxStalls") == 5

        queue.set_dlq_config({"autoRetry": True, "maxEntries": 500})
        dlq_config = queue.get_dlq_config()
        assert dlq_config.get("maxEntries") == 500, dlq_config


@test
def rate_limit_and_concurrency(server: Server) -> None:
    with Queue(unique_name("rate"), port=server.port) as queue:
        queue.set_global_rate_limit(100)
        queue.remove_global_rate_limit()
        queue.set_global_concurrency(2)
        queue.remove_global_concurrency()  # no exception = commands accepted


@test
def custom_rate_limit_duration_reaches_broker(server: Server) -> None:
    with Queue(unique_name("rate-window"), port=server.port) as queue:
        queue.set_global_rate_limit(7, 12345)
        response = queue.connection.call({"cmd": "GetQueueLimits", "queue": queue.name})
        limits = (response.get("data") or {}).get("limits") or {}
        rate_limit = limits.get("rateLimit") or {}
        assert rate_limit.get("duration") == 12345, rate_limit


@test
def global_concurrency_enforced(server: Server) -> None:
    with Queue(unique_name("gconc"), port=server.port) as queue:
        queue.set_global_concurrency(1)
        queue.add_bulk([{"name": "t", "data": {"i": i}} for i in range(4)])
        response = queue.connection.call(
            {"cmd": "PULLB", "queue": queue.name, "count": 4, "owner": "e2e", "lockTtl": 30000}
        )
        assert len(response.get("jobs") or []) <= 1, "global concurrency not enforced"
        queue.remove_global_concurrency()


@test
def webhooks_crud(server: Server) -> None:
    with Queue(unique_name("wh"), port=server.port) as queue:
        webhook_url = "https://example.com/bunqueue-hook"
        webhook_id = queue.add_webhook(webhook_url, ["job.completed"], queue=queue.name)
        hooks = queue.list_webhooks()
        assert any(h.get("url") == webhook_url for h in hooks), hooks
        assert webhook_id, "webhook id returned"
        queue.set_webhook_enabled(webhook_id, False)
        hooks = queue.list_webhooks()
        disabled = next(h for h in hooks if h.get("id") == webhook_id)
        assert disabled.get("enabled") is False, disabled
        queue.remove_webhook(webhook_id)
        hooks = queue.list_webhooks()
        assert not any(h.get("id") == webhook_id for h in hooks)


@test
def stats_metrics_queues(server: Server) -> None:
    with Queue(unique_name("mon"), port=server.port) as queue:
        queue.add("t", {})
        stats = queue.get_stats()
        assert "waiting" in stats and "uptime" in stats, stats
        metrics = queue.get_metrics()
        assert "totalPushed" in metrics, metrics
        queues = queue.list_queues()
        names = [q.get("name") if isinstance(q, dict) else q for q in queues]
        assert queue.name in names, queues
